diff --git a/src/Backends/DRMBackend.cpp b/src/Backends/DRMBackend.cpp
index dfd6c4a0..f6930486 100644
--- a/src/Backends/DRMBackend.cpp
+++ b/src/Backends/DRMBackend.cpp
@@ -3425,6 +3425,8 @@ namespace gamescope
 				return false;
 			}
 
+			vulkan_init_icc_profile();
+
 			return init_drm( &g_DRM, g_nPreferredOutputWidth, g_nPreferredOutputHeight, g_nNestedRefresh );
 		}
 
@@ -3515,6 +3517,7 @@ namespace gamescope
 
 			bNeedsFullComposite |= !!(g_uCompositeDebug & CompositeDebugFlag::Heatmap);
 			bNeedsFullComposite |= rotationShaderOrientation != GAMESCOPE_PANEL_ORIENTATION_0;
+			bNeedsFullComposite |= g_bIccProfileEnabled;
 
 			bool bDoComposite = true;
 			if ( !bNeedsFullComposite && !bWantsPartialComposite )
diff --git a/src/main.cpp b/src/main.cpp
index 7ca79cb9..f5bd564e 100644
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -93,6 +93,7 @@ const struct option *gamescope_options = (struct option[]){
 	{ "generate-drm-mode", required_argument, nullptr, 0 },
 	{ "immediate-flips", no_argument, nullptr, 0 },
 	{ "use-rotation-shader", no_argument, nullptr, 0 },
+	{ "icc-profile", required_argument, nullptr, 0 },
 	{ "framerate-limit", required_argument, nullptr, 0 },
 
 	// openvr options
@@ -231,6 +232,7 @@ const char usage[] =
 	"  --generate-drm-mode            DRM mode generation algorithm (cvt, fixed)\n"
 	"  --immediate-flips              Enable immediate flips, may result in tearing\n"
 	"  --use-rotation-shader          use shader to rotate output\n"
+	"  --icc-profile                  ICC color profile file to use\n"
 	"\n"
 #if HAVE_OPENVR
 	"VR mode options:\n"
@@ -304,6 +306,9 @@ bool g_bOutputHDREnabled = false;
 bool g_bRotationShaderRequested = false;
 bool g_bRotationShaderEnabled = false;
 
+bool g_bIccProfileEnabled = false;
+const char *g_sIccProfileFileName = nullptr;
+
 bool g_bFullscreen = false;
 bool g_bForceRelativeMouse = false;
 
@@ -811,6 +816,8 @@ int main(int argc, char **argv)
 					g_preferDeviceID = deviceID;
 				} else if (strcmp(opt_name, "use-rotation-shader") == 0) {
 					g_bRotationShaderRequested = true;
+				} else if (strcmp(opt_name, "icc-profile") == 0) {
+					g_sIccProfileFileName = optarg;
 				} else if (strcmp(opt_name, "immediate-flips") == 0) {
 					cv_tearing_enabled = true;
 				} else if (strcmp(opt_name, "force-grab-cursor") == 0) {
diff --git a/src/main.hpp b/src/main.hpp
index 3ed23421..ac55fa7e 100644
--- a/src/main.hpp
+++ b/src/main.hpp
@@ -25,6 +25,9 @@ extern bool g_bForceInternal;
 extern bool g_bRotationShaderRequested;
 extern bool g_bRotationShaderEnabled;
 
+extern bool g_bIccProfileEnabled;
+extern const char *g_sIccProfileFileName;
+
 extern bool g_bFullscreen;
 
 extern bool g_bGrabbed;
diff --git a/src/meson.build b/src/meson.build
index 8512bf57..d3c207cc 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -28,7 +28,7 @@ wlroots_dep = dependency(
   'wlroots-0.18',
   version: ['>= 0.18.0', '< 0.19.0'],
   fallback: 'wlroots',
-  default_options: ['default_library=static', 'examples=false', 'xwayland=enabled', 'backends=libinput', 'renderers=[]', 'allocators=[]', 'session=enabled'],
+  default_options: ['default_library=static', 'examples=false', 'xwayland=enabled', 'color-management=enabled', 'backends=libinput', 'renderers=[]', 'allocators=[]', 'session=enabled'],
 )
 
 displayinfo_dep = dependency(
@@ -92,7 +92,7 @@ reshade_include = include_directories([
 
 sol2_include = include_directories(['../thirdparty'])
 
-required_wlroots_features = ['xwayland']
+required_wlroots_features = ['xwayland', 'color_management']
 
 src = [
   'Backends/HeadlessBackend.cpp',
diff --git a/src/rendervulkan.cpp b/src/rendervulkan.cpp
index 81f46c9e..da7760cc 100644
--- a/src/rendervulkan.cpp
+++ b/src/rendervulkan.cpp
@@ -30,6 +30,8 @@
 #endif
 #include "wlr_begin.hpp"
 #include <wlr/render/drm_format_set.h>
+#include <wlr/render/color.h>
+#include <render/color.h>
 #include "wlr_end.hpp"
 
 #include "rendervulkan.hpp"
@@ -3580,6 +3582,135 @@ bool vulkan_init( VkInstance instance, VkSurfaceKHR surface )
 	return true;
 }
 
+/* Taken from swaywm */
+static bool read_file_into_buf(const char *path, void **buf, size_t *size) {
+	/* Why not use fopen/fread directly? glibc will successfully open directories,
+	 * not just files, and supports seeking on them. Instead, we directly
+	 * work with file descriptors and use the more consistent open/fstat/read. */
+	int fd = open(path, O_RDONLY | O_NOCTTY | O_CLOEXEC);
+	if (fd == -1) {
+		return false;
+	}
+	char *b = NULL;
+	struct stat info;
+	size_t nread = 0;
+	off_t s = 0;
+	if (fstat(fd, &info) == -1) {
+		goto fail;
+	}
+	// only regular files, to avoid issues with e.g. opening pipes
+	if (!S_ISREG(info.st_mode)) {
+		goto fail;
+	}
+	s = info.st_size;
+	if (s <= 0) {
+		goto fail;
+	}
+	b = (char *)calloc(1, s);
+	if (!b) {
+		goto fail;
+	}
+	while (nread < (size_t)s) {
+		size_t to_read = (size_t)s - nread;
+		ssize_t r = read(fd, b + nread, to_read);
+		if ((r == -1 && errno != EINTR) || r == 0) {
+			goto fail;
+		}
+		nread += (size_t)r;
+	}
+	close(fd);
+	*buf = b;
+	*size = (size_t)s;
+	return true; // success
+fail:
+	free(b);
+	close(fd);
+	return false;
+}
+
+static gamescope::Rc<CVulkanTexture> icc_create_3d_lut_image(const struct wlr_color_transform_lut3d *lut_3d) {
+	size_t dim_len = lut_3d->dim_len;
+	auto lut3d = vulkan_create_3d_lut(dim_len, dim_len, dim_len);
+
+	size_t lut3d_size = dim_len * dim_len * dim_len * sizeof(uint16_t) * 4;
+	auto [base_dst, base_offset] = g_device.uploadBufferData(lut3d_size);
+
+	uint16_t *dst = (uint16_t *)base_dst;
+	for (size_t b_index = 0; b_index < dim_len; b_index++) {
+		for (size_t g_index = 0; g_index < dim_len; g_index++) {
+			for (size_t r_index = 0; r_index < dim_len; r_index++) {
+				size_t sample_index = r_index + dim_len * g_index + dim_len * dim_len * b_index;
+				size_t src_offset = 3 * sample_index;
+				size_t dst_offset = 4 * sample_index;
+				dst[dst_offset] = quantize_lut_value_16bit(lut_3d->lut_3d[src_offset]);
+				dst[dst_offset + 1] = quantize_lut_value_16bit(lut_3d->lut_3d[src_offset + 1]);
+				dst[dst_offset + 2] = quantize_lut_value_16bit(lut_3d->lut_3d[src_offset + 2]);
+				dst[dst_offset + 3] = 0;
+			}
+		}
+	}
+
+	auto cmdBuffer = g_device.commandBuffer();
+	cmdBuffer->copyBufferToImage(g_device.uploadBuffer(), base_offset, 0, lut3d);
+	g_device.submit(std::move(cmdBuffer));
+	g_device.waitIdle();
+
+	return lut3d;
+}
+
+static gamescope::Rc<CVulkanTexture> g_icc_color_transform_tex;
+bool vulkan_init_icc_profile()
+{
+	if (!g_sIccProfileFileName) {
+		return false;
+	}
+
+	if (g_icc_color_transform_tex) {
+		g_bIccProfileEnabled = true;
+		return true;
+	}
+	g_bIccProfileEnabled = false;
+
+	void *data = NULL;
+	size_t size = 0;
+	struct wlr_color_transform *color_transform = nullptr;
+
+	if (!read_file_into_buf(g_sIccProfileFileName, &data, &size)) {
+		vk_log.errorf( "Failed to read ICC profile from file %s", g_sIccProfileFileName );
+		goto fail;
+	}
+
+	color_transform = wlr_color_transform_init_linear_to_icc(data, size);
+	if (!color_transform) {
+		vk_log.errorf( "Failed to initialize ICC profile from file %s", g_sIccProfileFileName );
+		goto fail;
+	}
+
+	if (color_transform->type != COLOR_TRANSFORM_LUT_3D) {
+		vk_log.errorf( "Unsupported ICC profile type from wlroots" );
+		goto fail;
+	}
+
+	g_icc_color_transform_tex = icc_create_3d_lut_image(&color_transform->lut3d);
+	if (!g_icc_color_transform_tex) {
+		vk_log.errorf( "Failed to create 3D LUT image for ICC profile" );
+		goto fail;
+	}
+
+	g_bIccProfileEnabled = true;
+
+fail:
+	if (data) {
+		free(data);
+		data = nullptr;
+	}
+	if (color_transform) {
+		wlr_color_transform_unref(color_transform);
+		color_transform = nullptr;
+	}
+	return g_bIccProfileEnabled;
+}
+
 gamescope::OwningRc<CVulkanTexture> vulkan_create_texture_from_dmabuf( struct wlr_dmabuf_attributes *pDMA, gamescope::OwningRc<gamescope::IBackendFb> pBackendFb )
 {
 	gamescope::OwningRc<CVulkanTexture> pTex = new CVulkanTexture();
@@ -4035,7 +4166,10 @@ std::optional<uint64_t> vulkan_composite( struct FrameInfo_t *frameInfo, gamesco
 	auto cmdBuffer = pInCommandBuffer ? std::move( pInCommandBuffer ) : g_device.commandBuffer();
 
 	for (uint32_t i = 0; i < EOTF_Count; i++)
-		cmdBuffer->bindColorMgmtLuts(i, frameInfo->shaperLut[i], frameInfo->lut3D[i]);
+		if ( g_bIccProfileEnabled && !pOutputOverride )
+			cmdBuffer->bindColorMgmtLuts(i, nullptr, g_icc_color_transform_tex);
+		else
+			cmdBuffer->bindColorMgmtLuts(i, frameInfo->shaperLut[i], frameInfo->lut3D[i]);
 
 	if ( frameInfo->useFSRLayer0 )
 	{
diff --git a/src/rendervulkan.hpp b/src/rendervulkan.hpp
index a2b6735c..a909d90b 100644
--- a/src/rendervulkan.hpp
+++ b/src/rendervulkan.hpp
@@ -404,6 +404,7 @@ namespace CompositeDebugFlag
 VkInstance vulkan_get_instance(void);
 bool vulkan_init(VkInstance instance, VkSurfaceKHR surface);
 bool vulkan_init_formats(void);
+bool vulkan_init_icc_profile(void);
 bool vulkan_make_output();
 
 gamescope::OwningRc<CVulkanTexture> vulkan_create_texture_from_dmabuf( struct wlr_dmabuf_attributes *pDMA, gamescope::OwningRc<gamescope::IBackendFb> pBackendFb );
diff --git a/src/shaders/composite.h b/src/shaders/composite.h
index 961379b0..e08f610d 100644
--- a/src/shaders/composite.h
+++ b/src/shaders/composite.h
@@ -104,11 +104,14 @@ vec3 apply_layer_color_mgmt(vec3 color, uint layer, uint colorspace) {
     //
     // We also need to do degamma here for non-linear views to blend in linear space.
     // ie. PQ -> PQ would need us to manually do bilinear here.
-    bool lut3d_enabled = textureQueryLevels(s_shaperLut[plane_eotf]) != 0;
+    bool lut3d_enabled = textureQueryLevels(s_lut3D[plane_eotf]) != 0;
     if (lut3d_enabled)
     {
-        color = colorspace_plane_shaper_tf(color, colorspace);
-        color = perform_1dlut(color, s_shaperLut[plane_eotf]);
+        bool shaper_enabled = textureQueryLevels(s_shaperLut[plane_eotf]) != 0;
+        if (shaper_enabled) {
+            color = colorspace_plane_shaper_tf(color, colorspace);
+            color = perform_1dlut(color, s_shaperLut[plane_eotf]);
+        }
         color = perform_3dlut(color, s_lut3D[plane_eotf]);
         color = colorspace_blend_tf(color, c_output_eotf);
     }
Submodule subprojects/wlroots contains modified content
diff --git a/subprojects/wlroots/render/color_lcms2.c b/subprojects/wlroots/render/color_lcms2.c
index fae4fb6a..7c2460a6 100644
--- a/subprojects/wlroots/render/color_lcms2.c
+++ b/subprojects/wlroots/render/color_lcms2.c
@@ -45,13 +45,14 @@ struct wlr_color_transform *wlr_color_transform_init_linear_to_icc(
 		goto out_icc_profile;
 	}
 
+	cmsToneCurve **vcgtTag = (cmsToneCurve **)cmsReadTag(icc_profile, cmsSigVcgtTag);
 	cmsToneCurve *linear_tf[] = {
 		linear_tone_curve,
 		linear_tone_curve,
 		linear_tone_curve,
 	};
 	cmsHPROFILE srgb_profile = cmsCreateRGBProfileTHR(ctx, &srgb_whitepoint,
-		&srgb_primaries, linear_tf);
+		&srgb_primaries, vcgtTag && *vcgtTag ? vcgtTag : linear_tf);
 	if (srgb_profile == NULL) {
 		wlr_log(WLR_ERROR, "cmsCreateRGBProfileTHR failed");
 		goto out_linear_tone_curve;
